File: android/src/main/kotlin/africa/ejara/trustdart/coins/TRX.kt
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.toLong
import com.google.protobuf.ByteString
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Tron

class TRX : Coin("TRX", CoinType.TRON) {

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val cmd = txData["cmd"] as String
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val txHash: String?
        when (cmd) {
            "TRC20" -> {
                val trc20Contract = Tron.TransferTRC20Contract.newBuilder()
                    .setOwnerAddress(txData["ownerAddress"] as String)
                    .setContractAddress(txData["contractAddress"] as String)
                    .setToAddress(txData["toAddress"] as String)
                    .setAmount(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["amount"] as String))))

                val blockHeader = Tron.BlockHeader.newBuilder()
                    .setTimestamp(txData["blockTime"]!!.toLong())
                    .setTxTrieRoot(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["txTrieRoot"] as String))))
                    .setParentHash(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["parentHash"] as String))))
                    .setNumber(txData["number"]!!.toLong())
                    .setWitnessAddress(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["witnessAddress"] as String))))
                    .setVersion(txData["version"] as Int)
                    .build()

                val transaction = Tron.Transaction.newBuilder()
                    .setTransferTrc20Contract(trc20Contract)
                    .setTimestamp(txData["timestamp"]!!.toLong())
                    .setFeeLimit(txData["feeLimit"]!!.toLong())
                    .setBlockHeader(blockHeader)
                    .build()

                val signingInput = Tron.SigningInput.newBuilder()
                    .setTransaction(transaction)
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))

                val output =
                    AnySigner.sign(signingInput.build(), coinType, Tron.SigningOutput.parser())
                txHash = output.json
            }
            "TRC10" -> {
                val trc10Contract = Tron.TransferAssetContract.newBuilder()
                    .setOwnerAddress(txData["ownerAddress"] as String)
                    .setAssetName(txData["assetName"] as String)
                    .setToAddress(txData["toAddress"] as String)
                    .setAmount(txData["amount"]!!.toLong())

                val blockHeader = Tron.BlockHeader.newBuilder()
                    .setTimestamp(txData["blockTime"]!!.toLong())
                    .setTxTrieRoot(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["txTrieRoot"] as String))))
                    .setParentHash(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["parentHash"] as String))))
                    .setNumber(txData["number"]!!.toLong())
                    .setWitnessAddress(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["witnessAddress"] as String))))
                    .setVersion(txData["version"] as Int)
                    .build()

                val transaction = Tron.Transaction.newBuilder()
                    .setTransferAsset(trc10Contract)
                    .setTimestamp(txData["timestamp"]!!.toLong())
                    .setBlockHeader(blockHeader)
                    .build()

                val signingInput = Tron.SigningInput.newBuilder()
                    .setTransaction(transaction)
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))

                val output =
                    AnySigner.sign(signingInput.build(), coinType, Tron.SigningOutput.parser())
                txHash = output.json
            }
            "TRX" -> {
                val transfer = Tron.TransferContract.newBuilder()
                    .setOwnerAddress(txData["ownerAddress"] as String)
                    .setToAddress(txData["toAddress"] as String)
                    .setAmount(txData["amount"]!!.toLong())

                val blockHeader = Tron.BlockHeader.newBuilder()
                    .setTimestamp(txData["blockTime"]!!.toLong())
                    .setTxTrieRoot(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["txTrieRoot"] as String))))
                    .setParentHash(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["parentHash"] as String))))
                    .setNumber(txData["number"]!!.toLong())
                    .setWitnessAddress(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["witnessAddress"] as String))))
                    .setVersion(txData["version"] as Int)
                    .build()

                val transaction = Tron.Transaction.newBuilder()
                    .setTimestamp(txData["timestamp"]!!.toLong())
                    .setFeeLimit(txData["feeLimit"]!!.toLong())
                    .setTransfer(transfer)
                    .setBlockHeader(blockHeader)
                    .build()

                val signingInput = Tron.SigningInput.newBuilder()
                    .setTransaction(transaction)
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))

                val output =
                    AnySigner.sign(signingInput.build(), coinType, Tron.SigningOutput.parser())
                txHash = output.json
            }
            "FREEZE" -> {
                val freezeContract = Tron.FreezeBalanceContract.newBuilder()
                    .setOwnerAddress(txData["ownerAddress"] as String)
                    .setResource(txData["resource"] as String)
                    .setFrozenDuration(txData["frozenDuration"]!!.toLong())
                    .setFrozenBalance(txData["frozenBalance"]!!.toLong())

                val blockHeader = Tron.BlockHeader.newBuilder()
                    .setTimestamp(txData["blockTime"]!!.toLong())
                    .setTxTrieRoot(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["txTrieRoot"] as String))))
                    .setParentHash(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["parentHash"] as String))))
                    .setNumber(txData["number"]!!.toLong())
                    .setWitnessAddress(ByteString.copyFrom(Numeric.hexStringToByteArray((txData["witnessAddress"] as String))))
                    .setVersion(txData["version"] as Int)
                    .build()

                val transaction = Tron.Transaction.newBuilder()
                    .setTimestamp(txData["timestamp"]!!.toLong())
                    .setFreezeBalance(freezeContract)
                    .setBlockHeader(blockHeader)
                    .build()

                val signingInput = Tron.SigningInput.newBuilder()
                    .setTransaction(transaction)
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))

                val output =
                    AnySigner.sign(signingInput.build(), coinType, Tron.SigningOutput.parser())
                txHash = output.json
            }
            "CONTRACT" -> {
                txHash = null
            }
            else -> txHash = null
        }
        return txHash
    }

}

File: ios/trustdart/Sources/trustdart/coins/TRX.swift
//
//  TRX.swift
//  trustdart
//
//  Created by Jay on 3/29/22.
//

import WalletCore

class TRX: Coin  {
    init() {
        super.init(name: "TRX", coinType: .tron)
    }
    
    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        let cmd = txData["cmd"] as! String
        var txHash: String?
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)?.getKey(coin: self.coinType, derivationPath: path)
        if privateKey != nil {
            switch cmd {
            case "TRC20":
                let contract = TronTransferTRC20Contract.with {
                    $0.ownerAddress = txData["ownerAddress"] as! String
                    $0.toAddress = txData["toAddress"] as! String
                    $0.contractAddress = txData["contractAddress"] as! String
                    $0.amount = Data(hexString: txData["amount"] as! String)!
                }
                
                let input = TronSigningInput.with {
                    $0.transaction = TronTransaction.with {
                        $0.transferTrc20Contract = contract
                        $0.feeLimit = txData["feeLimit"] as! Int64
                        $0.timestamp = txData["timestamp"] as! Int64
                        $0.blockHeader = TronBlockHeader.with {
                            $0.timestamp = txData["blockTime"] as! Int64
                            $0.number = txData["number"] as! Int64
                            $0.version = txData["version"] as! Int32
                            $0.txTrieRoot = Data(hexString: txData["txTrieRoot"] as! String)!
                            $0.parentHash = Data(hexString: txData["parentHash"] as! String)!
                            $0.witnessAddress = Data(hexString: txData["witnessAddress"] as! String)!
                        }
                    }
                    $0.privateKey = privateKey!.data
                }
                let output: TronSigningOutput = AnySigner.sign(input: input, coin: self.coinType)
                txHash = output.json
            case "TRC10":
                let transferAsset = TronTransferAssetContract.with {
                    $0.ownerAddress = txData["ownerAddress"] as! String
                    $0.toAddress = txData["toAddress"] as! String
                    $0.amount = txData["amount"] as! Int64
                    $0.assetName = txData["assetName"] as! String
                }
                let input = TronSigningInput.with {
                    $0.transaction = TronTransaction.with {
                        $0.transferAsset = transferAsset
                        $0.timestamp = txData["timestamp"] as! Int64
                        $0.blockHeader = TronBlockHeader.with {
                            $0.timestamp = txData["blockTime"] as! Int64
                            $0.number = txData["number"] as! Int64
                            $0.version = txData["version"] as! Int32
                            $0.txTrieRoot = Data(hexString: txData["txTrieRoot"] as! String)!
                            $0.parentHash = Data(hexString: txData["parentHash"] as! String)!
                            $0.witnessAddress = Data(hexString: txData["witnessAddress"] as! String)!
                        }
                    }
                    $0.privateKey = privateKey!.data
                }
                let output: TronSigningOutput = AnySigner.sign(input: input, coin: self.coinType)
                txHash = output.json
            case "TRX":
                let transfer = TronTransferContract.with {
                    $0.ownerAddress = txData["ownerAddress"] as! String
                    $0.toAddress = txData["toAddress"] as! String
                    $0.amount = txData["amount"] as! Int64
                }
                let input = TronSigningInput.with {
                    $0.transaction = TronTransaction.with {
                        $0.transfer = transfer
                        $0.feeLimit = txData["feeLimit"] as! Int64
                        $0.timestamp = txData["timestamp"] as! Int64
                        $0.blockHeader = TronBlockHeader.with {
                            $0.timestamp = txData["blockTime"] as! Int64
                            $0.number = txData["number"] as! Int64
                            $0.version = txData["version"] as! Int32
                            $0.txTrieRoot = Data(hexString: txData["txTrieRoot"] as! String)!
                            $0.parentHash = Data(hexString: txData["parentHash"] as! String)!
                            $0.witnessAddress = Data(hexString: txData["witnessAddress"] as! String)!
                        }
                    }
                    $0.privateKey = privateKey!.data
                }
                let output: TronSigningOutput = AnySigner.sign(input: input, coin: self.coinType)
                txHash = output.json
            case "CONTRACT":
                txHash = ""
            case "FREEZE":
                let contract = TronFreezeBalanceContract.with {
                    $0.frozenBalance = txData["frozenBalance"] as! Int64
                    $0.frozenDuration = txData["frozenDuration"] as! Int64
                    $0.ownerAddress = txData["ownerAddress"] as! String
                    $0.resource = txData["resource"] as! String
                }
                let input = TronSigningInput.with {
                    $0.transaction = TronTransaction.with {
                        $0.freezeBalance = contract
                        $0.timestamp = txData["timestamp"] as! Int64
                        $0.blockHeader = TronBlockHeader.with {
                            $0.timestamp = txData["blockTime"] as! Int64
                            $0.number = txData["number"] as! Int64
                            $0.version = txData["version"] as! Int32
                            $0.txTrieRoot = Data(hexString: txData["txTrieRoot"] as! String)!
                            $0.parentHash = Data(hexString: txData["parentHash"] as! String)!
                            $0.witnessAddress = Data(hexString: txData["witnessAddress"] as! String)!
                        }
                    }
                    $0.privateKey = privateKey!.data
                }
                let output: TronSigningOutput = AnySigner.sign(input: input, coin: self.coinType)
                txHash = output.json
            default:
                txHash = nil
            }
            return txHash
        }
        return nil
    }
}
